### DISCUSSION TITLE
How to define a container_element for a ui created from a background process / async process

### DISCUSSION THREAD

**Discussion by amorphousWaste** [zauberzeug/nicegui#3464] [Q&A] [answered]
### Question

Hello. I'm trying to use NiceGui to create a command console for a Twitch bot built around async operations. Using the existing design pattern I use for the rest of the bot, I cannot figure out how to properly create and run the UI. I've tried looking at the examples and reading through similar issues, but I haven't found an answer.

Here is a minimal example:
```python
import asyncio
from nicegui import ui

class Console():
    async def init(self) -> None:
        ui.label('Test')
        ui.run()

async def init() -> None:
    await Console().init()

if __name__ in ["__main__", "__mp_main__"]:
    asyncio.run(init())
```
and this is the error:
```
RuntimeError: The current slot cannot be determined because the slot stack for this task is empty.
This may happen if you try to create UI from a background task.
To fix this, enter the target slot explicitly using `with container_element:`.
```
Any help would be greatly appreciated. Thank you.

**Comment by rodja** [comment, marked as answer]
NiceGUI is starting it's own asyncio loop via FastAPI/Uvicorn so you simply write:

```py
from nicegui import ui, app

class Console():
    async def init(self) -> None:
        ui.label('Test')

async def init() -> None:
    await Console().init()

if __name__ in ["__main__", "__mp_main__"]:
    app.on_startup(init)
    ui.run()
```

**Reply by amorphousWaste** [reply]
Hello. Thank you for this. It definitely helps, but I think I made my example too minimal, as I don't think this will quite work with my setup. This is a better view of how the bot currently works:
```python
import asyncio

# bot imports

from nicegui import app, ui

class Console():
    def __init__(self, bot) -> None:
        self.bot = bot

    async def init(self) -> None:
        ui.label('Test')

async def init() -> None:
    authorization = await Auth().init()
    bot = await TwitchBot().init(authorization)
    message_receiver = ChatReceiver(chat_connection)
    tasks = []
    tasks.extend(
        [
            asyncio.create_task(bot.run()),
            asyncio.create_task(message_receiver.run()),
        ]
    )
    await asyncio.gather(*tasks)

if __name__ in ["__main__", "__mp_main__"]:
    asyncio.run(init())
```
The UI needs to communicate with the bot so I need to pass a bot instance into the UI. If I try to run the NiceGUI app in the `init()`, it errors because I'm trying to run an event loop from an event loop. I'm not sure where I can actually initialize the UI from. Is it possible to include the UI in the `asyncio.gather`? Apologies again for not being clear enough. Thank you.

**Reply by rodja** [reply]
Still, you should be able to use `ui.run` instead of `asyncio.run` and start `init` via `app.on_startup`.

**Reply by amorphousWaste** [reply]
That will work. Thank you very much for the help.
